Add scripts and actions for creating Windows containers for Windows 10 and 11.

wmaxey opened the pull request:
This makes 2 images for each compiler and CUDA entry in the matrix of:

2017, 2019, and 2022.
CUDA 11.1 and 12.1.

See packages here for results:
https://github.com/wmaxey?tab=packages&repo_name=cuda-windows-images

wmaxey commented on .github/workflows/trigger_windows.yml:
```diff
@@ -0,0 +1,75 @@
+      matrix:
+        runner: [windows-2019, windows-2022]
+        cuda_version: ["12.1", "11.1"]
+        edition: [windows-2019, windows-2022]
```
Runner/edition can be different in the future. We'll likely refactor this before a new edition of GHA runners comes out.

trxcllnt reviewed (approved):
LGTM. Should this be integrated into the main release job, or is it fine if we manually trigger releases separately?

wmaxey commented:
> LGTM. Should this be integrated into the main release job, or is it fine if we manually trigger releases separately?

I think manually is fine, or at the least your choice. Pushing the tags to dockerhub wasn't included in these changes.

trxcllnt commented:
Ok, we'll probably want to do that.

jrhemstad commented:
@miscco

miscco commented:
@trxcllnt I have made some minor modifications to the PR. With those I was able to locally build the images and run our CI scripts inside the container.

wmaxey commented:
> @trxcllnt I have made some minor modifications to the PR. With those I was able to locally build the images and run our CI scripts inside the container.

I unfortunately rebased and force pushed over everything. I've updated to 12.2 and fixed the PR thing so it actually runs jobs.

wmaxey commented:
Can publish images now only when desired by launching manually.

trxcllnt commented on .github/workflows/dispatch-windows.yml:
```diff
@@ -0,0 +1,80 @@
+      fail-fast: false
+      matrix:
+        runner: [windows-2019, windows-2022]
+        cuda_version: ["12.1", "11.1"]
```
?
```suggestion
        cuda_version: ["12.2", "11.1"]
```

trxcllnt commented on scripts/windows/make-image-manifest.ps1:
```diff
@@ -0,0 +1,38 @@
+    $image_name_2019="$(.\generate-image-name.ps1 -clVersion $clVersion -cudaVersion $cudaVersion -edition "windows2019" -repo $repo)"
+    $image_name_2022="$(.\generate-image-name.ps1 -clVersion $clVersion -cudaVersion $cudaVersion -edition "windows2022" -repo $repo)"
+
+    $manifest_name="${repo}:cuda${cudaVersion}-cl${clVersion}"
```
```suggestion
    $manifest_name="${repo}:${prefix}cuda${cudaVersion}-cl${clVersion}"
```

trxcllnt commented on scripts/windows/make-image-manifest.ps1:
```diff
@@ -0,0 +1,38 @@
+    $cudaVersion="latest",
+    [Parameter(Mandatory=$false)]
+    [string]
+    $repo="local"
```
```suggestion
    $repo="local",
    [Parameter(Mandatory=$false)]
    [string]
    $prefix=""
```

trxcllnt commented on .github/actions/create-and-push-manifest/action.ps1:
```diff
@@ -0,0 +1,21 @@
+    $cudaVersion="latest",
+    [Parameter(Mandatory=$true)]
+    [string]
+    $repo
```
```suggestion
    $repo,
    [Parameter(Mandatory=$false)]
    [string]
    $prefix=""
```

trxcllnt commented on .github/actions/create-and-push-manifest/action.ps1:
```diff
@@ -0,0 +1,21 @@
+$clVerArray = $vsVerToCompilers[$msvcVersion]
+
+foreach ($cl in $clVerArray) {
+    .\scripts\windows\make-image-manifest -clVersion $cl -cudaVersion $cudaVersion -repo $repo
```
```suggestion
    .\scripts\windows\make-image-manifest -clVersion $cl -cudaVersion $cudaVersion -repo $repo --prefix $prefix
```

trxcllnt commented on .github/actions/create-and-push-manifest/action.yml:
```diff
@@ -0,0 +1,28 @@
+        .github\actions\create-and-push-manifest\action.ps1 `
+          -cudaVersion ${{ inputs.cuda_version }}           `
+          -msvcVersion ${{ inputs.vs_version }}             `
+          -repo ${{ inputs.repo }}
```
Not sure about powershell syntax, but something like this?
```suggestion
          -repo ${{ inputs.repo }}
          -prefix "$(git describe --abbrev=0 --tags | sed 's/[a-zA-Z]//g' | cut -d '.' -f -2)-"
```

trxcllnt commented on scripts/windows/build-windows-image.ps1:
```diff
@@ -0,0 +1,80 @@
+
+    $vsVer = $vsYearToVer[$vsCompilersToYear[$clVersion]]
+    # Override defaults in .env.
+    $ENV:IMAGE_NAME="$(.\generate-image-name.ps1 -clVersion $clVersion -cudaVersion $cudaVersion -edition $edition -repo $repo)"
```
```suggestion
    $ENV:IMAGE_NAME="$(.\generate-image-name.ps1 -clVersion $clVersion -cudaVersion $cudaVersion -edition $edition -repo $repo -prefix $prefix)"
```

trxcllnt commented on scripts/windows/build-windows-image.ps1:
```diff
@@ -0,0 +1,80 @@
+    $isolation="hyperv",
+    [Parameter(Mandatory=$false)]
+    [string]
+    $repo="local"
```
```suggestion
    $repo="local",
    [Parameter(Mandatory=$false)]
    [string]
    $prefix=""
```

trxcllnt commented on .github/actions/build-windows-image/action.ps1:
```diff
@@ -0,0 +1,27 @@
+$clVerArray = $vsVerToCompilers[$msvcVersion]
+
+foreach ($cl in $clVerArray) {
+    .\scripts\windows\build-windows-image.ps1 -clVersion $cl -isolation $isolation -cudaVersion $cudaVersion -edition $edition -repo $repo
```
```suggestion
    .\scripts\windows\build-windows-image.ps1 -clVersion $cl -isolation $isolation -cudaVersion $cudaVersion -edition $edition -repo $repo -prefix $prefix
```

trxcllnt commented on .github/actions/build-windows-image/action.ps1:
```diff
@@ -0,0 +1,27 @@
+    $isolation="hyperv",
+    [Parameter(Mandatory=$true)]
+    [string]
+    $repo
```
```suggestion
    $repo,
    [Parameter(Mandatory=$false)]
    [string]
    $prefix=""
```

trxcllnt commented on .github/actions/build-windows-image/action.yml:
```diff
@@ -0,0 +1,38 @@
+          -msvcVersion ${{ inputs.vs_version }}        `
+          -edition ${{ inputs.win_edition }}           `
+          -isolation ${{ inputs.isolation }}           `
+          -repo ${{ inputs.repo }}
```
Not sure about powershell syntax, but something like this?
```suggestion
          -repo ${{ inputs.repo }}
          -prefix "$(git describe --abbrev=0 --tags | sed 's/[a-zA-Z]//g' | cut -d '.' -f -2)-"
```

miscco commented:
@trxcllnt how is the state of the PR, can it be merged?

trxcllnt commented:
/ok to test
